From cb9f7402bb132580764bc01e4ba30cea9ecf1b53 Mon Sep 17 00:00:00 2001
From: Ramya Gnanasekar <quic_rgnanase@quicinc.com>
Date: Mon, 12 Dec 2022 19:13:38 +0530
Subject: [PATCH] hostapd: Set duplicate beacon bit in HE oper IE

When APUT operates in non-psc channels, duplicate beacons
will be sent in psc channels. Hence duplicate beacon field
to be updated in the same scenario.

Signed-off-by: Ramya Gnanasekar <quic_rgnanase@quicinc.com>

--- a/src/ap/ieee802_11_he.c
+++ b/src/ap/ieee802_11_he.c
@@ -268,6 +268,11 @@ u8 * hostapd_eid_he_operation(struct hos
 		control |= hapd->iconf->he_6ghz_reg_pwr_type <<
 			HE_6GHZ_OPER_INFO_CTRL_REG_INFO_SHIFT;
 
+		if (center_idx_to_bw_6ghz(seg0) &&
+		    !is_6ghz_psc_frequency(ieee80211_chan_to_freq(NULL,
+				    hapd->iconf->op_class, hapd->iconf->channel)))
+			control |= HE_6GHZ_OPER_INFO_CTRL_DUP_BEACON;
+
 		*pos++ = control;
 
 		/* Channel Center Freq Seg0/Seg1 */
